packages/lit-html/src/lib/lit-html.ts

```diff
@@ -213,6 +243,13 @@ const walker = d.createTreeWalker(d);
 // Interfaces and type aliases can be interleaved freely.
 //
 
+export abstract class Directive {
+  abstract render(...props: Array<unknown>): unknown;
+  update(_part: Part, props: Array<unknown>): unknown {
+    return this.render(...props);
```

Comment:
It's a little odd that `update` returns a value; it means that when doing custom update logic you might need to `return super.update(part, props)` (as opposed to just calling `super.update(part, props)`, which isn't super intuitive given the name.

All things equal, it would seem better to keep semantics similar to LitElement by doing this:
```js
update(part, props) {
  part.__setValue(this.render(...props));
}
```
And having `__commitDirective` just call `this.__directive.update(this, value.values))`

Reply:
I see the pattern used in tests is for update to call/return render directly rather than super'ing.

Oh, and I guess the current factoring of attribute parts doesn't lend well to this... Let's discuss.

Reply:
Chatted offline. I think side-effectful update is a bit tricky due to the inconsistent __setValue() API.